packages/pg-v5/lib/psql.js

```diff
@@ -1,107 +1,247 @@
+// is a child process of the Heroku CLI node process.
+const trapAndForwardSignalsToChildProcess = (childProcess) => {
+  const signalsToTrap = ['SIGINT']
+  const signalTraps = signalsToTrap.map((signal) => {
+    process.removeAllListeners(signal);
+    const listener = () => kill(childProcess, signal)
+    process.on(signal, listener)
+    return [signal, listener]
+  });
+
+  // restores the built-in node ctrl+c and other handlers
+  const cleanup = () => {
+    signalTraps.forEach(([signal, listener]) => {
+      process.removeListener(signal, listener)
     })
-    handlePsqlError(reject, psql)
-  })
+  }
+
+  return cleanup
 }
 
-function psqlInteractive (dbEnv, prompt) {
-  const { spawn } = require('child_process')
-  return new Promise((resolve, reject) => {
-    let psqlArgs = ['--set', `PROMPT1=${prompt}`, '--set', `PROMPT2=${prompt}`]
-    let psqlHistoryPath = process.env.HEROKU_PSQL_HISTORY
-    if (psqlHistoryPath) {
-      const fs = require('fs')
-      const path = require('path')
-      if (fs.existsSync(psqlHistoryPath) && fs.statSync(psqlHistoryPath).isDirectory()) {
-        let appLogFile = `${psqlHistoryPath}/${prompt.split(':')[0]}`
-        debug('Logging psql history to %s', appLogFile)
-        psqlArgs = psqlArgs.concat(['--set', `HISTFILE=${appLogFile}`])
-      } else if (fs.existsSync(path.dirname(psqlHistoryPath))) {
-        debug('Logging psql history to %s', psqlHistoryPath)
-        psqlArgs = psqlArgs.concat(['--set', `HISTFILE=${psqlHistoryPath}`])
-      } else {
-        const cli = require('heroku-cli-util')
-        cli.warn(`HEROKU_PSQL_HISTORY is set but is not a valid path (${psqlHistoryPath})`)
+async function runWithTunnel (db, tunnelConfig, options) {
+  const tunnel = await Tunnel.connect(db, tunnelConfig)
+  debug('after create tunnel')
+
+  const psql = execPSQL(options)
+  const cleanupSignalTraps = trapAndForwardSignalsToChildProcess(psql)
+
+  try {
+    debug('waiting for psql or tunnel to exit')
+    // wait for either psql or tunnel to exit;
+    // the important bit is that we ensure both processes are
+    // always cleaned up in the `finally` block below
+    await Promise.race([
+      waitForPSQLExit(psql),
+      tunnel.waitForClose()
+    ])
+  } catch (err) {
+    debug('wait for psql or tunnel error', err)
+    throw err
+  } finally {
+    debug('begin tunnel cleanup')
+    cleanupSignalTraps()
+    tunnel.close()
+    kill(psql, 'SIGKILL')
+    debug('end tunnel cleanup')
+  }
+}
+
+// a small wrapper around tunnel-ssh
+// so that other code doesn't have to worry about
+// whether there is or is not a tunnel
+class Tunnel {
+  constructor (bastionTunnel) {
+    this.bastionTunnel = bastionTunnel
+    this.events = new EventEmitter()
+  }
+
+  async waitForClose () {
+    if (this.bastionTunnel) {
+      try {
+        debug('wait for tunnel close')
+        await once(this.bastionTunnel, 'close')
```

Comment:
[events.once](https://nodejs.org/docs/latest-v14.x/api/events.html#events_events_once_emitter_name) will return a promise and resolve with the value of the event. If, while it is waiting for the event, an `error` event occurs, it will reject the promise. That's why you don't see the `.on('error')` event handler carried over in the new version.